polars/polars-arrow/src/kernels/rolling/nulls/min_max.rs

```diff
@@ -1,460 +1,256 @@
-use arrow::bitmap::utils::{count_zeros, ZipValidityIter};
 use nulls;
 use nulls::{rolling_apply_agg_window, RollingAggWindowNulls};
 
 use super::*;
 
-pub fn is_reverse_sorted_max_nulls<T: NativeType + PartialOrd + IsFloat>(
-    values: &[T],
-    validity: &Bitmap,
-) -> bool {
-    let mut current_max = None;
-    for opt_v in ZipValidityIter::new(values.iter(), validity.iter()) {
-        match (current_max, opt_v) {
-            // do nothing
-            (None, None) => {}
-            (None, Some(v)) => current_max = Some(*v),
-            (Some(current), Some(val)) => {
-                match compare_fn_nan_min(&current, val) {
-                    Ordering::Greater => {
-                        current_max = Some(*val);
-                    }
-                    // allowed
-                    Ordering::Equal => {}
-                    // not sorted
-                    Ordering::Less => return false,
-                }
-            }
-            (Some(_current), None) => {}
-        }
-    }
-
-    true
-}
-
-pub struct SortedMinMax<'a, T: NativeType> {
+pub struct MinMaxWindow<'a, T: NativeType + PartialOrd + IsFloat, const R: i8> {
     slice: &'a [T],
     validity: &'a Bitmap,
+    m: Option<T>,
+    m_idx: usize,
+    sorted_to: usize,
     last_start: usize,
     last_end: usize,
     null_count: usize,
 }
 
-impl<'a, T: NativeType> SortedMinMax<'a, T> {
-    fn count_nulls(&self, start: usize, end: usize) -> usize {
-        let (bytes, offset, _) = self.validity.as_slice();
-        count_zeros(bytes, offset + start, end - start)
-    }
-}
-
-impl<'a, T: NativeType> RollingAggWindowNulls<'a, T> for SortedMinMax<'a, T> {
-    unsafe fn new(
-        slice: &'a [T],
-        validity: &'a Bitmap,
-        start: usize,
-        end: usize,
-        _params: DynArgs,
-    ) -> Self {
-        let mut out = Self {
-            slice,
-            validity,
-            last_start: start,
-            last_end: end,
-            null_count: 0,
-        };
-        let nulls = out.count_nulls(start, end);
-        out.null_count = nulls;
-        out
-    }
-
-    unsafe fn update(&mut self, start: usize, end: usize) -> Option<T> {
-        self.null_count -= self.count_nulls(self.last_start, start);
-        self.null_count += self.count_nulls(self.last_end, end);
-
-        self.last_start = start;
-        self.last_end = end;
-
-        // return first non null
-        for idx in start..end {
-            let valid = self.validity.get_bit_unchecked(idx);
-
-            if valid {
-                return Some(*self.slice.get_unchecked(idx));
+// R should be an Ordering but isn't because non-primitive const generics aren't considered stable
+// as of this writing. Passing comparison information this way is faster than having a field inside
+// the struct so we live with it.
```

Comment:
Could you explain to me what you wanted to do without the `R` trick? 

I believe it should be solvable with an intermediate trait.


What was your initial solution that didn't work?

Reply:
I tried it a few different ways. One was to pass a comparison function in to the constructor and call that dynamically. That's what the previous version of the code did.
The other was to take the Ordering passed in and use that or it's `reverse()` to determine where `NaN` goes. Both were slower, though the second was a bit better.

The current solution seems roughly on par with macros.

If there's a way to do it by adding more traits I guess we can do that. In general, I'm loath to add such structure and complexity if it isn't absolutely needed. But this is pretty deep internals, it's localized, and I care much more about how the external facing parts look for end-users. Besides, if it's more verbose I'll get more GitHubPoints :-)